packages/portfolio-deploy/src/contract-control.js

```diff
@@ -0,0 +1,308 @@
+          installation,
+          issuers,
+          privateArgsOverrides,
+        });
+      },
+
+      /** @param {string} bundleId */
+      async upgrade(bundleId) {
+        mustMatch(bundleId, iface0.upgrade);
+        const { name, revoked, kit } = this.state;
+        trace(name, 'upgrade', bundleId);
+        !revoked || Fail`revoked`;
+        if (!kit) throw Fail`${q(name)}: no StartedInstanceKit`;
+        const { privateArgs } = kit;
+        const result = await E(kit.adminFacet).upgradeContract(
+          bundleId,
+          privateArgs,
+        );
+        trace(name, 'upgrade result', result);
+        return result;
+      },
+
+      /**
+       * @param {object} opts
+       * @param {string} [opts.target] boardId to confirm is current
+       * @param {string} [opts.message] for termination error
+       * @param {boolean} [opts.revoke] neuter this object
+       */
+      async terminate(opts = {}) {
+        mustMatch(opts, iface0.terminate);
+        const { name, kit, revoked } = this.state;
+        trace(name, 'terminate', kit?.adminFacet, opts);
+        !revoked || Fail`revoked`;
+        const { target, message, revoke } = opts;
+        if (!kit) {
+          if (revoke) {
+            trace(name, 'revoked');
+            this.state.revoked = true;
+            return;
+          }
+          throw Fail`${q(name)}: no StartedInstanceKit`;
+        }
+
+        await null;
+        if (target) {
+          const current = await E(svcs.board).getId(kit.instance);
+          assert.equal(current, target);
+        }
+
+        await E(kit.adminFacet).terminateContract(Error(message));
+        this.state.kit = undefined;
+        if (revoke) {
+          trace(name, 'revoked');
+          this.state.revoked = true;
+        }
+        const { agoricNamesAdmin } = svcs;
+        const instanceAdmin = E(agoricNamesAdmin).lookupAdmin('instance');
+        await E(instanceAdmin).delete(name);
+      },
+
+      /** @param {Record<string, string[]>} parentToChildren */
+      async pruneChainStorage(parentToChildren) {
+        mustMatch(parentToChildren, iface0.pruneChainStorage);
+        const { name, storageNode, revoked } = this.state;
+        trace(name, 'pruneChainStorage', Object.keys(parentToChildren).length);
+        !revoked || Fail`revoked`;
+
+        /** @param {string[]} path */
+        const makePathNode = path => {
+          /** @type {Promise<StorageNode>} */
+          // @ts-expect-error Remote/E integration incomplete
+          let node = Promise.resolve(storageNode);
+          for (const segment of path) {
+            node = E(node).makeChildNode(segment);
+          }
+          return node;
+        };
+
+        await null;
+        let qty = 0;
+        for (const [parent, children] of Object.entries(parentToChildren)) {
+          const segments = parent
+            .replace(/^(published|orchtest)\./, '')
+            .split('.');
+          assert.equal(segments[0], name);
```

Comment:
I don't understand this hard coding here. Why don't we simply `getPath` of the root `storageNode` stored in `state`, and strip that?

Reply:
that hadn't occurred to me.

I'm not certain that all the mock storage nodes support `getPath`, so I'm hesitant to change it. Maybe if I need another trip thru ci